zboxcore/sdk/writemarker_mutex.go

```diff
@@ -0,0 +1,260 @@
+
+		now := time.Now()
+		body.Set("request_time", strconv.FormatInt(now.Unix(), 10))
+		buf := bytes.NewBufferString(body.Encode())
+
+		for {
+
+			// M locks are acquired, it is safe to commit write
+			if n >= M {
+				return nil
+			}
+
+			// No more blobber, but n < M
+			if i > T && n < M {
+				//fails, release all locks
+				m.Unlock(ctx, connectionID) //nolint: errcheck
+				break
+			}
+
+			result, err := m.lockOne(ctx, buf, urls[i-1])
+
+			// current blobber fails or is down, try next blobber
+			if err != nil {
+				// fails on current blobber
+				i++
+				continue
+			}
+
+			// it is locked by other session, wait and retry
+			if result.Status == WMLockStatusPending {
+				logger.Logger.Info("WriteMarkerLock is pending, wait and retry")
+				time.Sleep(1 * time.Second)
+				continue
+			} else if result.Status == WMLockStatusOK {
+				// locked on current blobber, count it and go to next blobber
+				i++
+				n++
+			}
+		}
+
+		if retry >= 2 {
+			return constants.ErrNotLockedWritMarker
+		}
+
+		time.Sleep(1 * time.Second)
+	}
+}
+
+// lockOne acquire WriteMarker lock from a blobber
+func (m *WriteMarkerMutex) lockOne(ctx context.Context, buf *bytes.Buffer, url string) (*WMLockResult, error) {
+	transport := &http.Transport{
+		Dial: (&net.Dialer{
+			Timeout: resty.DefaultDialTimeout,
+		}).Dial,
+		TLSHandshakeTimeout: resty.DefaultDialTimeout,
+	}
+
+	result := &WMLockResult{}
+
+	r := resty.New(transport, func(req *http.Request, resp *http.Response, respBody []byte, cf context.CancelFunc, err error) error {
+		if err != nil {
+			return err
+		}
+
+		err = json.Unmarshal(respBody, result)
+
+		if err != nil {
+			return err
+		}
+
+		return nil
+	})
+
+	r.DoPost(ctx, buf, url)
+
+	err := r.Wait()
+	if len(err) > 0 {
+		return nil, err[0]
+	}
+
+	return result, nil
+}
+
+// Unlock release WriteMarker lock on blobbers
+func (m *WriteMarkerMutex) Unlock(ctx context.Context, sessionID string) error {
+
+	if m == nil || m.allocationObj == nil {
+		return errors.Throw(constants.ErrInvalidParameter, "allocationObj")
+	}
+
+	T := len(m.allocationObj.Blobbers)
+
+	// no blobbers, it is unnecessary to release locks
+	if T == 0 {
+		return nil
+	}
+
+	urls := make([]string, T)
+
+	for i, b := range m.allocationObj.Blobbers {
+		urls[i] = strings.TrimRight(b.Baseurl, "/") + blobber.EndpointWriteMarkerLock + m.allocationObj.Tx
+	}
+
+	transport := &http.Transport{
+		Dial: (&net.Dialer{
+			Timeout: resty.DefaultDialTimeout,
+		}).Dial,
+		TLSHandshakeTimeout: resty.DefaultDialTimeout,
+	}
+
+	r := resty.New(transport, func(req *http.Request, resp *http.Response, respBody []byte, cf context.CancelFunc, err error) error {
+		return err
+	})
+
+	r.DoDelete(ctx, urls...)
+
+	errs := r.Wait()
+
+	if len(errs) == 0 {
+		return nil
+	}
+
+	msgList := make([]string, 0, len(errs))
+	for _, err := range errs {
+		msgList = append(msgList, err.Error())
+	}
+
+	return errors.Throw(constants.ErrNotUnlockedWritMarker, msgList...)
```

Comment:
Will those blobbers who didn't receive lock requests return error here? if yes, then we need to separate those errors from others, or we should avoid sending unlock request to those blobbers that we didn't send lock requests to.

Reply:
fixed it on urls array.